pkg/aspect/quotasManager.go

```diff
@@ -58,94 +74,129 @@ func (m *quotasManager) NewAspect(c *config.Combined, a adapter.Builder, env ada
+	// checking via the Check API, which doesn't generate a QuotaMethodArgs
+	if !ok {
+		qma = &QuotaMethodArgs{
+			Quota:           "RequestCount",
+			Amount:          1,
+			DeduplicationID: strconv.FormatInt(atomic.AddInt64(&w.manager.dedupCounter, 1), 16),
+			BestEffort:      false,
 		}
 	}
 
-	// TODO: for now we don't support dedup semantics
-	dedupID := strconv.FormatInt(atomic.AddInt64(&w.manager.dedupCounter, 1), 16)
+	info, ok := w.metadata[qma.Quota]
+	if !ok {
+		return &Output{Code: code.Code_INVALID_ARGUMENT}, fmt.Errorf("unknown quota '%s'", qma.Quota)
+	}
 
-	for i, d := range w.descriptors {
-		qa := prepQuotaArgs(attrs, &d, dedupID, labels)
+	labels, err := evalAll(info.labels, attrs, mapper)
+	if err != nil {
+		return &Output{Code: code.Code_INTERNAL}, fmt.Errorf("failed to evaluate labels for quota '%s' with err: %s", qma.Quota, err)
+	}
 
-		// TODO: for now we only support Alloc semantics, no AllocBestEffort or ReleaseBestEffort
-		if amount, err := w.aspect.Alloc(qa); err != nil || amount <= 0 {
-			// something went wrong, return any allocated quota
-			for j := i; j >= 0; j-- {
-				qa := prepQuotaArgs(attrs, &w.descriptors[j], dedupID, labels)
-				_, _ = w.aspect.ReleaseBestEffort(qa)
-			}
+	qa := adapter.QuotaArgs{
+		Definition:      info.definition,
+		Labels:          labels,
+		QuotaAmount:     qma.Amount,
+		DeduplicationID: qma.DeduplicationID,
+	}
 
-			if err != nil {
-				return &Output{Code: code.Code_INTERNAL}, err
-			}
+	var amount int64
 
-			return &Output{Code: code.Code_RESOURCE_EXHAUSTED}, nil
-		}
+	if qma.BestEffort {
+		amount, err = w.aspect.AllocBestEffort(qa)
+	} else {
+		amount, err = w.aspect.Alloc(qa)
 	}
 
-	return &Output{Code: code.Code_OK}, nil
-}
+	if err != nil {
+		return &Output{Code: code.Code_INTERNAL}, err
+	}
 
-func prepQuotaArgs(attrs attribute.Bag, d *dpb.QuotaDescriptor,
-	dedupID string, labels map[string]interface{}) adapter.QuotaArgs {
+	if amount == 0 {
+		return &Output{Code: code.Code_RESOURCE_EXHAUSTED}, nil
+	}
 
-	// TODO: for now, assume no one passed in the amount
-	amount := int64(1)
+	// TODO: need to return the allocated amount somehow in the Quota API's QuotaResponse message
 
-	qa := adapter.QuotaArgs{
-		Name:            d.Name,
-		Labels:          make(map[string]interface{}),
-		QuotaAmount:     amount,
-		DeduplicationID: dedupID,
-	}
-
-	/*
-		for _, a := range d.Labels {
-			if val, ok := labels[a]; ok {
-				qa.Labels[a] = val
-				continue
-			}
-			if val, found := attribute.Value(attrs, a); found {
-				qa.Labels[a] = val
-			}
-		}
-	*/
-	return qa
+	return &Output{Code: code.Code_OK}, nil
 }
 
 func (w *quotasWrapper) Close() error {
 	return w.aspect.Close()
 }
+
+func findQuota(quotas []*aconfig.QuotasParams_Quota, name string) *aconfig.QuotasParams_Quota {
+	for _, q := range quotas {
+		if q.Descriptor_ == name {
+			return q
+		}
+	}
+	return nil
+}
+
+func quotaDescToDef(desc *dpb.QuotaDescriptor) (*adapter.QuotaDefinition, error) {
```

Comment:
similar to above: fwiw, `definitionFromProto` would more closely align with the other methods like this, including `ptypes.DurationFromProto` which is used within this method.